### ISSUE TITLE
Severe performance problems

### ISSUE THREAD

**Issue by kordianbruck** [engelsystem/engelsystem#300]
So at 33C3 we have real problems with performance. Can we do some brainstorming on how we can improve this, so that being an angel is not such a pain in the ass?

For example:

1. Use some caching: APC(u), memcached for heavy queries
2. Reduce the number of filter attributes in long running queries (Some of them are unneeded, as many people are only basic angels)
3. Add indexes? I can only see two indexes in the upgrade.sql - really? Are there only selects by PKs/FKs otherwise?
4. Do not run queries in loops. Precompute the needed info then during rendering do not fetch additional data from the DB
5. Use a template system like twig to encourage the separation of logic from views
6. Reduce the number of HTML elements by using absolute positioning
7. Filter the shifts on the client side

Let me know, how I can help to get this going, by supplying a better server or setting up a database cluster with replication.

**Comment by kiliankoe** [comment]
WOᴙKꙄ ꟻOᴙ MƎ

😛

**Comment by kordianbruck** [comment]
So for example, after looking at the code for a bit, the ```NeededAngelTypes_by_shift``` function is called in loop during the rendering of the timetable: https://github.com/engelsystem/engelsystem/blob/master/includes/model/NeededAngelTypes_model.php#L58

The "fix" (https://github.com/engelsystem/engelsystem/commit/b05b235cf821ec2d9fc84779d09db71dc1bf7797) is not really great. We could save several data points directly into the record like the number of open spots, so this does not need to be recalculated each time.

Caching probably be the most significant thing, that would make this render a lot faster. I would need to know if APC is available before I start working on a PR.

**Comment by mortzu** [comment]
We have to fix some server configuration issues first (Switch from Apache to nginx, switch from mod_php to php-fpm, maybe switch to php7?). I hope we can do this after the 33c3 system is needed so we can fix code performance issues soon.

**Comment by kordianbruck** [comment]
Im happy to help with that migration as well, as i have recently just done this on my infrastructure. Let me know, if that is possible.

**Comment by claudehohl** [comment]
What do you think about the idea of making use of the web sql browser database? Instead of 2000 users performing complex filtering-queries on the poor mysql server, 2000 browsers performing the queries, fetching only initial data and update from time to time.

**Comment by kordianbruck** [comment]
@claudehohl absolutely a great idea. Not only should we filter the shifts on the client with javascript, but we should rewrite the calendar to use absolute positioned shifts instead of the current layout of mechanism of rendering 15 minute ticks.

**Comment by claudehohl** [comment]
Great! I'm getting excited! :)

![angelsystemwebsql](https://cloud.githubusercontent.com/assets/496245/21578839/ad2ae522-cf90-11e6-8d48-52264fa4fc01.png)

 Just put together a little proof of concept and familiarized myself with the angelsystem. The database looks simple. We only need to fetch shifts and nicknames (as far as I can see?). Rest of the user data doesn't leave the server.

The next steps would be to transfer the necessary data/tables to the browserdb and run queries on it.

Then comes the development of the (client-side) renderer, which surely takes some time to implement because it's quite extensive.

Shifts* and User* table need to record an *updated* field. There needs to be a client-side fetcher that fetches those tables where updated > last_update and make the updates. That check can happen every five minutes or to keep the load chilled.

Those are roughly my ideas that I developed on my trainride home from 33c3. I'm not sure when I will have time for this all, but I guess the deadline is 34c3? I'll create a branch in my fork and start working.

**Comment by MyIgel** [comment]
First we should decide on #269 and refactor the database acces to use PDO
Atm. the engelsystem is 0% PHP7.1 compatible

**Comment by msquare** [comment]
Please note, that the actual version of the system meets the performance requirements for 2.500 angels easy (like no more cpu usage or load on the server).

**Comment by claudehohl** [comment]
So, what was the issue then that caused minutes of waiting time for the Shifts view?

**Comment by kordianbruck** [comment]
The calculation of the free or full shifts. The last time I checked this was removed from the system all together, so one only sees all shifts, not only the ones that need help

**Comment by davidspinat** [comment]
@claudehohl we could just use the localstorage instead http://caniuse.com/#search=localstorage

**Comment by claudehohl** [comment]
@davidspinat Yea, and in the meantime I found a nifty library: http://alasql.org/ - uses IndexedDB as storage backend while still providing SQL to the client, and it works on all browsers. Still it has it's issues; for example primary key constraints don't work. But hey, it's a start.

**Comment by claudehohl** [comment]
My progress so far:

![angelwebsql](https://cloud.githubusercontent.com/assets/496245/21606133/c8826218-d1ad-11e6-8c18-1413945189cc.gif)

* I made an export for all neccessary tables, results in a single json
* JavaScript fetches this json, imports it into the browser db
* Tested with 6'000 shifts
* I can run complex SQL queries with joins on it

Works for me.

**Comment by fabian-rump** [comment]
+1 for the idea of filtering on the client side. Not only to reduce performance impact on the database server (which btw should be totally manageable if the database and the queries are sufficiently optimized), but also to provide a more responsive way of filtering shifts. If the shifts could be provided via some sort of REST API, I could potentially build something in jQuery/Angular/whatever ...

**Comment by claudehohl** [comment]
Hold on, I'm almost done with the JS renderer. https://github.com/claudehohl/engelsystem/blob/websql/js/shifts.render.coffee

So far, I'm quite happy with the performance. Let's see if the performance allows for something more responsive (like refreshing the shift-calendar on checkmark-click). I'll find it out when I start working on filtering and complex (client-side) SQL queries in the next days.

https://github.com/claudehohl/engelsystem/commits/websql - to be continued.

**Comment by claudehohl** [comment]
So yea, I found some time again for developing, and I have reached a working, testable state.

## See it in action

This is how it looks: ![Angelsystem after update](http://f.scratchbook.ch/angelsystem-after-update.gif)

Note the instant updates. Should the rendering take too long (e.g. for large events), the filter button is automatically activated, and a progress bar is being displayed.

Rendering of colliding shifts works too (which currently doesn't in the legacy shiftplan). Also notice some improvements on the timelane: the current time is marked blue, and the time between 20:00 and 8:00 is rendered slightly darker to mark the night. The date is only displayed on day changes, and last but not least, there is a new datepicker.

Look how much network traffic is passing when I click and modify the filter settings: http://f.scratchbook.ch/angelsystem-network-traffic.gif - zero, none, nada. That means the server has more space to breathe. The intense computing tasks have been shifted to the browser. (This is filtering of a large event with 10k shifts).

All data only needs to be transferred initially. Should this process take too long (e.g. on Firefox or on Android devices), it can always be skipped, and the original shiftplan is being displayed: http://f.scratchbook.ch/angelsystem-synchronisation-and-graceful-degradation.gif

This is also the default behaviour for all non-js browsers. Afterwards, only incremental updates need to be transferred.

## Try it out for yourself

Small event: https://engelsystem.scratchbook.ch/
Large event: https://engelsystem-large.scratchbook.ch/

Logins for both systems are:
User: Engel1, Pw: 1
User: Engel2, Pw: 2
User: Engel3, Pw: 3
User: Engel4, Pw: 4

@msquare: Could you provide me some anonymized dump of the 33c3? I need to know how it performs with actual data. Or can you give me some Kennzahlen about the size of the tables, e.g. Shifts and NeededAngelTypes? How many rows are in there?

## Changes

Changes to the database:
* None.

New files:
* /js/* - coffeescript source files for this websql (shifts.js)
* public/js/shifts.js - the main, compiled everything.
* public/vendor/alasql.min.js - browser-sql library
* public/vendor/mustache.min.js - template-engine
* public/vendor/jquery.datetimepicker.css and
* public/vendor/jquery.datetimepicker.full.min.js - the new datepicker
* includes/pages/user_shifts_browser.php - fills the template with variables (currently only UID)

Changed files:
* config/config.default.php - added the option to enable websql or have it disabled.
* includes/controller/shifts_controller.php - renders the JSON
* includes/engelsystem_provider.php - added the browser view
* includes/model/Shifts_model.php - provides the data for the browserdb
* includes/pages/user_shifts.php - logic whether or not to enable websql (cookie-handling)
* public/css/theme0.css - added styles for nighttime rendering of the timelane
* public/index.php - added the new viewfunction
* templates/layout.html and
* templates/user_shifts_browser.html - include the required js- and css-files

You can see the full list of changes here: [Changed files for engelsystem:master..engelsystem:websql](https://github.com/engelsystem/engelsystem/compare/master...claudehohl:websql#files_bucket)

## Things to do

Things left to do for me until it's ready for a pull request:

* Translations
* Keep track of deleted and updated entries, and update the browserdb accordingly. This means db-changes to the angelsystem; i need to add an 'updated' column on all tables needed by the browserdb, and a new table "deleted entries".
* Rendering of: restricted angeltypes, when user != angeltype, edit+delete buttons for the admin (currently always displayed), and freeloaded angels.

I'ts another week of work, maybe less. After that, we are ready for 34C3!

**Comment by claudehohl** [comment]
Is my work appreciated at all? Or should I stop? I don't want to invest several more weeks, just to find out when it comes to the pull-request that it will be rejected... Give me some feedback please. A simple "yes, continue" is enough.

German version:

Wird meine Arbeit überhaupt geschätzt? Oder soll ich aufhören? Ich mag nicht noch ein paar Wochen Aufwand reinstecken, nur um dann zu sehen, dass mein Pull-Request abgelehnt oder genau so ignoriert wird wie meine Kommentare hier. Gebt mir mal ein kleines Feedback bitte, damit ich weiss, ob es sich lohnt. Ein kurzes "Ja, mach weiter" genügt.

**Comment by MyIgel** [comment]
Ich finde es gut, kommt aber wohl leider darauf an was @msquare davon hält? (Meinen WIP PR hat er bisher auch ignoriert...).
Solange eine Änderung mehr Performance bringt sollte sie aber mmn. gemerged werden ;)

**Comment by fabian-rump** [comment]
Given the situation at the last congress I can imagine that the congress orgas and shift coordinators are very interested in your work. I will probably see some of the involved people next week and will hopefully be able to have a chat with them about this :smile:

**Comment by msquare** [comment]
First of all: I'm not ignoring you all. I have simply very little time to maintain the engelsystem.

Regarding performance: After fixing the performance issues on 33c3 the server went idle again.

@claudehohl From a technical point of view your work is very impressive. The performance is about the same as the server rendered version. I think this is far to much work for resolving performance issues. And I fear sync issues and privacy issues and maintaining js code. But I would like to talk to you at 34c3 drinking some mate or tschunk to check if there is a way to give this feature a chance from my perspective.

I will also close this "master" issue. It's to big and includes several different topics which was/will be resolved in other issues.

**Comment by MyIgel** [comment]
@msquare @claudehohl I think we should organize a "Improve the Engelsystem" session or something similar at 34c3 ;)

**Comment by claudehohl** [comment]
Thanks for the response @msquare.

> I think this is far to much work for resolving performance issues.

Tell me about it! :D
But since I already invested many weeks, I did it anyway and finished it. :smiley_cat:

> Regarding performance: After fixing the performance issues on 33c3 the server went idle again.

Well, by disabling some filtering- and rendering-features... Note that in the gif above, users get the full-on calendar with all colors enabled. Rendering of collisions is disabled in the serverside shiftplan for example. Or is this a bug in the rendering and normal behaviour? Because in my version, the calendar looks as it's supposed to be.

> And I fear sync issues

Don't worry, I've taken care of that. Additional rows and a deletelog track every change and sync it back to the browserdb. In case of an emergency, I increase the versionstring and all browser-dbs get rebuilt from scratch.

> privacy issues

All I'm storing in the browser-db is data that is publicly available anyways. User-IDs, nicknames, room- and shiftdata. The rest of the data like emails never leave the server. Users need to be logged in to access this sync-data.

> maintaining js code. But I would like to talk to you at 34c3 drinking some mate or tschunk to check if there is a way to give this feature a chance from my perspective.

Yea let's sit together and have a drink! I'll introduce you to this whole thing step by step, how it works.

Main code resides in /js, and I maintain it. Should you choose to disable this feature at some point, you can safely do so. Legacy shiftview still works, nothing in the database is deleted, just some columns and a table added. They won't harm the system if they're not used. This feature is available only on browsers which support websql anyway, and it can be disabled in config.php entirely.

**Comment by MyIgel** [comment]
I think it looks nice!

Some things that I noticed:
* Please don't use $_*[...] variables!
* Maybe the whole shift fetch process should be done ove a clear dokumented api (some where there is a feature requrest for that an is planned after #343 and [Milestone 11)](https://github.com/engelsystem/engelsystem/milestone/11)
* As you are working on some frontend files (js & css), i think we should introduce a better build system for them (pack all together/minify and so on)

**Comment by claudehohl** [comment]
> $_*[...] variables!

I don't quite get it... is this JS/jQuery-related or PHP? Did you mean variables with CamelCase instead of under_score?

> Maybe the whole shift fetch process should be done ove a clear dokumented api

Yes. I tinkered something together because I needed it. But if there is or will be an API - even better!

Pack and minify frontend-files: Definitely! But this would've been out of scope for me, so I'll let this be part of another pullrequest.

**Comment by MyIgel** [comment]
I mean `$_GET`, please use `$request = request();` and then `$requrest->has('foo')` and `$request->input('foo')` (See [src/Http/Request.php](https://github.com/engelsystem/engelsystem/blob/master/src/Http/Request.php))

**Comment by claudehohl** [comment]
Ah, of course. Die Macht der Routine. I'm used to CodeIgniter where those values are automatically sanitized.

Fixed in https://github.com/engelsystem/engelsystem/pull/380/commits/50d147af7169a9e3a25aaf51269df676f9ef2398

**Comment by MyIgel** [comment]
Ping @claudehohl, @msquare when do you have time at the congress?

**Comment by msquare** [comment]
Now @heaven.

**Comment by msquare** [comment]
@claudehohl would you like to come to heaven?

**Comment by heaven** [comment]
@msquare now I am here :)

**Comment by msquare** [comment]
@heaven: Sorry, we didn't actually mean you, but a location at 34c3 (Chaos Communication Congress) called "heaven".
Nvm...

**Comment by claudehohl** [comment]
Ouch, I didn't check my email this day...

On a sidenote, I think the Angelsystem performs well and my contribution isn't needed, is it?

**Comment by msquare** [comment]
No problem. I really want to talk to you. Just show up in heaven and ask for me, if possible.
